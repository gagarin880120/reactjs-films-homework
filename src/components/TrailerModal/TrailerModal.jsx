import React, {useEffect, useState} from 'react';
import styles from './TrailerModal.module.scss';
import getTrailer from '../../services/services';

export default function TrailerModal(props) {
  const [trailerSource, setTrailerSource] = useState('loading');

  useEffect(() => {
    async function fetchData() {
      try {
        const data = await getTrailer(props.id);
        setTrailerSource(data.videos.results[0].key)
      } catch (error) {
        setTrailerSource(null)
      }
    }

    fetchData();
  }, [props.modalIsOpen])

  return (
    <div
      className={styles.modal}
      testid="modal"
      onClick={() => props.setModalIsOpen(false)}
    >
      {trailerSource === 'loading' ? null : trailerSource ?
        <iframe
          data-testid="video"
          width="800"
          height="450"
          src={`https://www.youtube.com/embed/${trailerSource}`}
          frameBorder="0"
          allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
        ></iframe> :
        <span data-testid="noTrailer" className={styles.noTrailer}>
          Trailer is not available
        </span>
      }
    </div>
  )
}
