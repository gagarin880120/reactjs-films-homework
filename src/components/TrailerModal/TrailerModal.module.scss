.modal {
  width: 100%;
  height: 100%;
  position: fixed;
  background-color: rgba(0, 0, 0, 0.7);
  top: 0;
  left: 0;
  justify-content: center;
  align-items: center;
  z-index: 10;
}

.noTrailer {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 54px;
  color: #fff;
  border: 3px solid #fff;
  border-radius: 10px;
  padding: 2%;
}
