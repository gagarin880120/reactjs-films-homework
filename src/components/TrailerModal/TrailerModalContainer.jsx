import React, { useEffect } from 'react';
import TrailerModal from './TrailerModal';
import { connect } from 'react-redux';
import { modalSelector, trailerSelector } from '../../redux/selectors';
import { setModal } from '../../redux/actions';

export function TrailerModalContainer({isModalOpen, closeModal, trailerURL}) {

  return (
    <TrailerModal
      isModalOpen={isModalOpen}
      closeModal={closeModal}
      trailerURL={trailerURL}
    />
  )
}

export const mapStateToProps = state => {
  return {
    isModalOpen: modalSelector(state),
    trailerURL: trailerSelector(state)
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    closeModal() {
      dispatch(setModal(false));
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(TrailerModalContainer);
