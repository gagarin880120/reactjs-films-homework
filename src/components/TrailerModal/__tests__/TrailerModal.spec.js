import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import TrailerModal from '../TrailerModal';
import {render, fireEvent} from '@testing-library/react';

describe('TrailerModal component', () => {
  test('TrailerModal component should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<TrailerModal />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  test('If modal is open should have display=flex', () => {
    const { getByTestId } = render(<TrailerModal modalIsOpen={true} />);
    const modal = getByTestId('modal');
    expect(modal.style.display).toBe('flex');
  });

  test('If trailer is loading should not render iframe element', () => {
    const { queryByTestId } = render(<TrailerModal isTrailerLoading={true} />);
    expect(queryByTestId('video')).toBeNull();
  });

  test('If trailerSource is, should render iframe element', () => {
    const { queryByTestId } = render(<TrailerModal trailerSource="path_to_video" />);
    expect(queryByTestId('video')).not.toBeNull();
  });

  test('testing onClick event', () => {
    const setIsOpen = jest.fn();
    const { getByTestId } = render(<TrailerModal setIsOpen={setIsOpen} />);
    const modal = getByTestId('modal');
    fireEvent.click(modal);
    expect(setIsOpen).toHaveBeenCalledTimes(1);
  });
});
