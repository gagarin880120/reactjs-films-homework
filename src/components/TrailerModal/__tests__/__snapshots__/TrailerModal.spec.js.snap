// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrailerModal component TrailerModal component should render without crashing 1`] = `
<div
  className="modal"
  data-testid="modal"
  onClick={[Function]}
  style={
    Object {
      "display": "none",
    }
  }
>
  <span
    className="noTrailer"
    data-testid="noTrailer"
  >
    Trailer is not available
  </span>
</div>
`;
